lib/protocol/bep_extensions.go

```diff
@@ -126,6 +127,72 @@ func (f FileInfo) IsEmpty() bool {
 	return f.Version.Counters == nil
 }
 
+// IsEqual checks that the two file infos represent the same actual file.
+// Permissions (config) and blocks (scanning) can be excluded from the comparison.
+// Any file info is not "equal", if it has different
+//  - type
+//  - deleted flag
+//  - invalid flag
+//  - permissions, unless they are ignored
+// A file is not "equal", if it has different
+//  - modification time
+//  - size
+//  - blocks, unless there are no blocks to compare (scanning)
+// A symlink is not "equal", if it has different
+//  - target
+// A directory does not have anything specific to check.
+func (f FileInfo) IsEqual(other FileInfo, ignorePerms bool, noBlocks bool) bool {
+	if f.IsEmpty() || other.IsEmpty() {
+		return f.IsEmpty() == other.IsEmpty()
```

Comment:
With this definition, two fileinfos can be "equal" if they both have no version, even if everything else is different. That's odd and might bite us in the future.